agent/src/main.rs

```diff
@@ -1573,15 +1572,49 @@ fn worker(
                             quota,
                         );
 
-                        // if regions need to be created, do that before apply_smf.
-                        let region_dataset = regions_dataset
+                        // If regions need to be created, do that before
+                        // apply_smf.
+                        let region_dataset = match regions_dataset
                             .ensure_child_dataset(
                                 &r.id.0,
                                 Some(reservation),
                                 Some(quota),
                                 &log,
-                            )
-                            .unwrap();
+                            ) {
+                            Ok(region_dataset) => region_dataset,
+                            Err(e) => {
+                                error!(
+                                    log,
+                                    "Dataset {} creation failed: {}",
+                                    &r.id.0,
+                                    e,
+                                );
+                                if let Err(e) = df.destroyed(&r.id) {
```

Comment:
Why not `df.fail` here?

Reply:
I had `df.fail()` there originally, but then the disk create saga unwinds and trys to delete the
disk, and we can't delete a failed disk.  I also had tried making it Tombstoned, and that made
the saga unhappy as well.  

The current stuff here does not require any changes on the Omicron side.  I think eventually we
may want to pass a specific message back to Omicron that we lack the space for this request.
Also in the work to do column is having Omicron do better accounting and not even try to allocate
a disk if we don't have space for it.